### PR TITLE
Avoid bisect command to get stuck

### PR DESCRIPTION
**Pull request by benoittgt**
Fix: https://github.com/rspec/rspec-core/issues/2637

---
From Palkan in benoittgt/rspec_repro_bisect_deadlock#1

First, I've tried to play with the number of specs which led to the
interesting conclusion: **the process hangs only at 1548+ specs**.

```diff
 RSpec.describe "a bunch of nothing" do
   (0...(ENV.fetch('N', 3000).to_i)).each do |t|
     it { expect(t).to eq t }
   end
 end
```

Try to run with `N=1547` and `N=1548`.

Seems suspicious, right?

Let's add `pry-byebug` to the equation (or Gemfile).

In order it to work we need to tweak our runner code a bit:

```diff
- $stdout = $stderr = @spec_output
+ # $stdout = $stderr = @spec_output
```

After a bit of `puts` debugging I localized the problem:
[`@channel.send`](https://github.com/rspec/rspec-core/blob/7b6b9c3/lib/rspec/core/bisect/fork_runner.rb#L122).

`Channel#send` calls `IO#write` here
https://github.com/rspec/rspec-core/blob/7b6b9c3/lib/rspec/core/bisect/utilities.rb#L41:

 ```ruby
 def send(message)
   packet = Marshal.dump(message)
   @write_io.write("#{packet.bytesize}\n#{packet}")
 end
 ```

 Do you know, what is the `packet.bytesize` for `N=1548`? It's **65548**.
 This number is very important: the pipe size is only **65536** on MacOS
 (see docs for [`IO#write_nonblock`](https://ruby-doc.org/core-2.6.3/IO.html#method-i-write_nonblock)
 for more).

 That makes `@write_io.write` hangs forever, because no one reads the
 buffer: we call `Channel#receive` only after `Process.waitpid(pid)`,
 thus waiting for the write operation to complete.

 -----------

 A basic proposal is to use WNOHANG. From waitpid doc:

 > If WNOHANG was specified in options and there were no children
 > in a waitable state, then waitid() returns 0 immediately (...)

 To validate this proposal on OSX we run just before running bisect:
 `lsof -n -P -r1 -c ruby | grep -e 'PIP' -e '===' -e 'COMMAND'`

 This will give us in loop the PIPE sizes of Ruby processes. Without our
 patch we see that quickly we hit 65536 bytes on two pipes, with the patch
 we keep pipes at the right size.

```
COMMAND PID    USER     FD   TYPE DEVICE                SIZE/OFF  NODE NAME
ruby    40134  benoit    3   PIPE 0xf3b025a6a6cd6005    16384     ->0xf3b025a6a6cd5045
ruby    40134  benoit    4   PIPE 0xf3b025a6a6cd5045    16384     ->0xf3b025a6a6cd6005
ruby    40134  benoit    5   PIPE 0xf3b025a6a6cd7805    16384     ->0xf3b025a6a6cd7145
ruby    40134  benoit    7   PIPE 0xf3b025a6a6cd7145    16384     ->0xf3b025a6a6cd7805
ruby    40134  benoit   10   PIPE 0xf3b025a6a6cd6fc5    16384     ->0xf3b025a6a6cd5a05
ruby    40134  benoit   11   PIPE 0xf3b025a6a6cd5a05    16384     ->0xf3b025a6a6cd6fc5
ruby    40144  benoit    3   PIPE 0xf3b025a6a6cd5d05    16384     ->0xf3b025a6a6cd5c45
ruby    40144  benoit    4   PIPE 0xf3b025a6a6cd5c45    16384     ->0xf3b025a6a6cd5d05
ruby    40144  benoit    5   PIPE 0xf3b025a6a6cd7085    16384     ->0xf3b025a6a6cd6785
ruby    40144  benoit    7   PIPE 0xf3b025a6a6cd6785    16384     ->0xf3b025a6a6cd7085
ruby    40144  benoit   10   PIPE 0xf3b025a6a6cd6fc5    16384     ->0xf3b025a6a6cd5a05
ruby    40144  benoit   11   PIPE 0xf3b025a6a6cd5a05    16384     ->0xf3b025a6a6cd6fc5
```

Improvements:
The bisect command request lot's of ram. The next step should be to
reduce that consumption.

_Thanks to Palkan, julik and Samuel Williams for the help. 🙇🏻‍♂️_

### CODE REVIEW

**Comment by pirj** [comment]
Just my 2c. Wondering why `waitpid` is needed at all? `read` on a pipe seems to be a blocking operation, and while the child process has not closed the channel, it won't return.
It might be a hasty conclusion, but it seems that `waitpid` is preventing the parent process from reading from that buffer and locks the child on a write operation.
I can't say I completely understand the semantics of `WNOHANG`, but it seems that we're relying on its behaviour of immediately returning anyway. However, we make ourselves dependent on the 'child in a waitable state', potentially risking a racing condition. There's slightly other description of it I've found which states:
> If status information is immediately available on an appropriate child process, waitpid() returns this information. Otherwise, waitpid() returns immediately with an error code indicating that the information was not available. In other words, WNOHANG checks child processes without causing the caller to be suspended.

With this in mind, do we really need to check that information that `waitpid` returns? We don't seem to use it.

Out-of-context reproduction:
```ruby
def experiment(n)
  rd, wr = IO.pipe
  pid = fork { rd.close; sleep 1; wr.write('a' * n); sleep 1; wr.close }
  Process.waitpid(pid)
  wr.close
  puts 'reading'
  puts rd.read.size
  puts 'done'
end
```

```ruby
> experiment(65536)
reading
65536
done
> experiment(65537)
```
the second experiment hangs forever.

However, if you comment out `Process.waitpid`, there is no difference in behaviour, except that `reading` is printed immediately, not together with read buffer size and `done`. Two seconds later read buffer size and `done` appear.

With this in mind, what do you think guys of removing `Process.waitpid` call altogether?

Great in-depth research by the way! 💯

**Comment by benoittgt** [comment]
Thanks a lot @pirj for this very interesting feedback. I think I thought about removing it at first but was not sure about unexpected behaviors.

I redo all my test and removing `Process.waitpid` seems to be a good idea! I will amend my commit

File: lib/rspec/core/bisect/fork_runner.rb [hunk 1]
```diff
@@ -92,7 +92,6 @@ def initialize(runner, channel)
 
           def dispatch_specs(run_descriptor)
             pid = fork { run_specs(run_descriptor) }
```
Comment (JonRowe) [inline lib/rspec/core/bisect/fork_runner.rb]: Need to not capture pid if this is the change we're making

Comment (benoittgt) [reply lib/rspec/core/bisect/fork_runner.rb]: Yep! Thanks

**Comment by JonRowe** [comment]
Are we happy this fixes the problem? Have we got a way to check for a regression to this in our tests?

**Comment by benoittgt** [comment]
> Are we happy this fixes the problem?

I am quite ok with the result even if I would love to test it on a linux and windows.

> Have we got a way to check for a regression to this in our tests?

One test we could have is to run bisect for a lot's of test. If you want I can look at this.

**Comment by JonRowe** [comment]
I think we need something...

**Comment by benoittgt** [comment]
Sorry for the delay.

I added a test on integration side. It was the easiest to implement in my case. I was able to reproduce the lock without my patch, then test pass with `waitpid` removed.

File: spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_ [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Deliberately named *.rb_ to avoid being loaded except when specified
+
+RSpec.describe "1000 tests" do
+  puts "Try to saturate the pipe in Bisect command"
+  (0..1000).each do |t|
```
Comment (pirj) [inline spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Shouldn't it be 1548+?

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: We run spec, that run a bisect command. When I did my test, I hit the cap  
with this amount of test. It pass with 922 occurrences, but not with 923.

With:
```ruby
  (0..923).each do |t|
    it { expect(t).to eq t }
  end
```
and `$ rspec --format doc` it get stuck here.

Capture d’écran 2019-12-06 à 16 58 00

Comment (JonRowe) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: I wonder if this is platform dependant and what we can do about that...

Comment (pirj) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Does it fail CI with 923?

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: I tried on a VM with Linux ElementaryOS. I have the same result. With 923 we get stuck.

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: I think we are good with the number 1000. Feel free to ask for more changes.

Comment (pirj) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: I would really feel safer if it failed on CI without the fix. Otherwise looks good.

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Good idea. So we have the issue on most ruby with 923 tests https://travis-ci.org/rspec/rspec-core/builds/623575109
Capture d’écran 2019-12-11 à 10 13 43

With 922 tests no stalled build https://travis-ci.org/rspec/rspec-core/builds/623575109

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Should we limit the test to OSX and Linux platform?

Comment (pirj) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Awesome!
You've linked the same (stalling) build, but I trust you on that one :D

I'd keep the test for all platforms, let's see if someone complains.

Comment (benoittgt) [reply spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_]: Oups sorry:

923 tests: https://travis-ci.org/rspec/rspec-core/builds/623575109
922 tests: https://travis-ci.org/rspec/rspec-core/builds/623579071

File: spec/integration/bisect_spec.rb [hunk 3]
```diff
@@ -32,5 +32,13 @@ def bisect(cli_args, expected_status=nil)
         expect(output).to include("Bisect failed!", "The example ordering is inconsistent")
       end
     end
+
+    context "when the bisect command is long" do
+      it 'does not hit pipe size limit and does not get stuck' do
+        output = bisect(%W[spec/rspec/core/resources/blocking_pipe_bisect_spec.rb_], 1)
+        puts output
```
Comment (JonRowe) [inline spec/integration/bisect_spec.rb]: We don't need the `puts output` committed 🙃

File: lib/rspec/core/bisect/fork_runner.rb [hunk 4]
```diff
@@ -91,8 +91,7 @@ def initialize(runner, channel)
           end
 
           def dispatch_specs(run_descriptor)
-            pid = fork { run_specs(run_descriptor) }
-            Process.waitpid(pid)
+            fork { run_specs(run_descriptor) }
```
Comment (JonRowe) [inline lib/rspec/core/bisect/fork_runner.rb]: Can we add a comment explaining why we don't need to wait here?

File: spec/integration/bisect_spec.rb [hunk 5]
```diff
@@ -32,5 +32,13 @@ def bisect(cli_args, expected_status=nil)
         expect(output).to include("Bisect failed!", "The example ordering is inconsistent")
       end
     end
+
+    context "when the bisect command is long" do
+      it 'does not hit pipe size limit and does not get stuck' do
```
Comment (JonRowe) [inline spec/integration/bisect_spec.rb]: Can we add a comment explaining the issue and linking here?

File: spec/integration/bisect_spec.rb (hunk 3 above)
Comment (benoittgt) [reply spec/integration/bisect_spec.rb]: Olalala. Sorry.

**Comment by benoittgt** [comment]
Thanks for the review. I am not about the clarity of my comment. Open for comments.

It was a pleasure to work on this subject.

File: spec/integration/bisect_spec.rb [hunk 6]
```diff
@@ -32,5 +32,17 @@ def bisect(cli_args, expected_status=nil)
         expect(output).to include("Bisect failed!", "The example ordering is inconsistent")
       end
     end
+
+    context "when the bisect command is long" do
+      # On OSX and Linux when we have many tests we could had pipe size issue. We did not empty the buffer
+      # and we ended up with a pipe that was full and it was blocking the main process. This made the bisect
+      # command stuck without further information. This test run blocking_pipe_bisect_spec.rb_ that
+      # will ensure that we are not blocked anymore in this kind of situation.
```
Comment (JonRowe) [inline spec/integration/bisect_spec.rb]: Can we rephrase this?

> On OSX and Linux a file descriptor limit meant that the bisect process got stuck at a certain limit. This test demonstrates that we can run large bisects above this limit (found to be <n> at time of commit). See:

File: lib/rspec/core/bisect/fork_runner.rb [hunk 7]
```diff
@@ -91,8 +91,11 @@ def initialize(runner, channel)
...
+            # We don't wait anymore that the subprocess is done with a waitpid
+            # because there is no need to. On OSX and Linux, with more than
+            # 1548 tests we were filling-up the pipe and saturating it. This
+            # ended up with stuck bisect command.
```
Comment (JonRowe) [inline lib/rspec/core/bisect/fork_runner.rb]: Personally I'd rephrase this:

> We don't use `Process.waitpid` here as it was causing bisects to block due to the file descriptor limit on OSX / Linux.

Or similar.

**Comment by benoittgt** [comment]
Thanks @JonRowe. It's much better and shorter with your proposal.

**Comment by JonRowe** [comment]
Happy for you to merge @benoittgt don't forget the change log :)

**Comment by jcabasc** [comment]
Hey guys @benoittgt is there a special way to run this command? I am running into the same issue, so wanted to double-check with you. By the way, which version is this one? Thanks.

**Comment by benoittgt** [comment]
This is the master version. We did not release a new version with the patch. You could point in your gemfile rspec-core to master.

**Comment by jcabasc** [comment]
Thanks man @benoittgt

**Comment by pirj** [comment]
@andrykonchin [brought up an interesting point](https://andrykonchin.github.io/rails/2019/12/25/deadlock-in-rspec.html) that by removing `Process.waitpid` we're producing zombie processes in the system.

A quick search showed that to mitigate this we can use [`Process.detach(pid)`](https://ruby-doc.org/core-2.2.2/Process.html#method-c-detach).

Maybe @ioquatix can advise here?

**Comment by JonRowe** [comment]
> @andrykonchin brought up an interesting point that by removing Process.waitpid we're producing zombie processes in the system.

This probably means that we have not fixed the problem, merely ignored it.

**Comment by benoittgt** [comment]
> @andrykonchin brought up an interesting point that by removing Process.waitpid we're producing zombie processes in the system.

Ouhao. Very nice article. 👏

This is a google translation but this is written:
> On the one hand, they solved the problem with deadlock , but on the other, they created a new one. Now the child process remains in the “zombie” state, since no one gets its completion status by a system call waitpid, which means leak of process descriptors. When the number of processes in the operating system reaches the limit, and it is not so big, it will not be possible to create a new child process.

Mentioned in [`waitpid` doc](https://linux.die.net/man/2/waitpid):
> A child that terminates, but has not been waited for becomes a "zombie". The kernel maintains a minimal set of information about the zombie process (PID, termination status, resource usage information) in order to allow the parent to later perform a wait to obtain information about the child. As long as a zombie is not removed from the system via a wait, it will consume a slot in the kernel process table, and if this table fills, it will not be possible to create further processes. If a parent process terminates, then its "zombie" children (if any) are adopted by init(8), which automatically performs a wait to remove the zombies. 

**So after few tests, we have probably zombie processes when the bisect commands run. It is not clean while the bisect command is running but it is cleaned after.**
**If your bisect command needs to run many step to isolate the failings specs suite and your machine doesn't have enough process available it will ended up using all your processes and the machine become unusable.**

**`detach` seems to be a good idea. I tested it on the code example below and did not hit any limitations.**

_________

I did this test, because I wanted to understand exactly what was happening and also try @pirj  proposal:

1. Create process and do not `waitpid` like `rspec-core` at the moment.

```ruby
@read_io, @write_io = IO.pipe

# write into pipe some data
def run_specs(i)
  packet = '*' * i
  @write_io.write("#{packet.bytesize}\n#{packet}")
end

count = 0
(65500..69500).each do |i|
  count += 1
  # create a child process
  pid = fork { run_specs(i) }
  puts "interval: #{i}, pid is: #{pid} and parent pid is: #{Process.ppid}, count: #{count}"

  # wait for its terminating
  # Process.waitpid(pid)
  sleep 0.1

  # read result
  packet_size = Integer(@read_io.gets)
  packet = @read_io.read(packet_size)

  puts "packet size: #{packet.size}"
end
```

2. Get max number of process on my mac for my user
```sh
$ sysctl kern.maxprocperuid
kern.maxprocperuid: 1418
```

3. Monitor process count for my user
```sh
while true; do pstree -u <username> | wc -l ; sleep 1; clear; done
```
You can also monitor zombie process count `while true; do ps axo pid=,stat= | awk '$2~/^Z/ { print }' | wc -l ; sleep 1; clear; done`

4. Run the ruby script. When we reach 1418 process the systems cannot create more process, it's problematic but to go there I was able to run create nearly 1000 process (that become zombie). They are cleared when the ruby script is cancel or over.
